src/suffix.c

```diff
@@ -124,55 +128,211 @@ void deleteSuffixTrie(Trie *trie, const char *str, uint32_t len) {
     TrieNode **children = __trieNode_children(node);
     for (int i = 0; i < node->numChildren; ++i) {
-      // printf("child %d ", i);
-      if (recursiveAdd(children[i], callback, ctx) != REDISMODULE_OK) {
+      if (recursiveAdd(children[i], sufCtx) != REDISMODULE_OK) {
         return REDISMODULE_ERR;
       }
     }
   }
   return REDISMODULE_OK;
 }
 
-void Suffix_IterateContains(TrieNode *n, const rune *str, size_t nstr, bool prefix,
-                            TrieSuffixCallback callback, void *ctx) {
-  if (prefix) {
+void Suffix_IterateContains(SuffixCtx *sufCtx) {
+  if (sufCtx->type == SUFFIX_TYPE_CONTAINS) {
     // get string from node and children
-    TrieNode *node = TrieNode_Get(n, str, nstr, 0, NULL);
+    TrieNode *node = TrieNode_Get(sufCtx->root, sufCtx->rune, sufCtx->runelen, 0, NULL);
     if (!node) {
       return;
     }
-    recursiveAdd(node, callback, ctx);
-  } else {
+    recursiveAdd(node, sufCtx);
+  } else if (sufCtx->type == SUFFIX_TYPE_SUFFIX) {
     // exact match. Get strings from a single node
-    TrieNode *node = TrieNode_Get(n, str, nstr, 1, NULL);
+    TrieNode *node = TrieNode_Get(sufCtx->root, sufCtx->rune, sufCtx->runelen, 1, NULL);
     suffixData *data = Suffix_GetData(node);
     if (data) {
-      processSuffixData(data, callback, ctx);
+      processSuffixData(data, sufCtx);
+    }
+  }
+}
+
+
+/***********************************************************************************
+*                                    Wildcard                                      *
+************************************************************************************/
+
+/* Breaks wildcard at '*'s and find the best token to get iterate the suffix trie  */
+int Suffix_ChooseToken(const char *str, size_t len, size_t *tokenIdx, size_t *tokenLen) {
+  int runner = 0;
+  int i = 0;
+  int init = 0;
+  while (i < len) {
+    if (str[i] != '*') {
+      tokenIdx[runner] = i;
+      init = 1;
+    }
+    while (i < len && str[i] != '*') {
+      ++i;
+    }
+    if (init) {
+      tokenLen[runner] = i - tokenIdx[runner];
+      ++runner;
+    }
+    while (str[i] == '*') {
+      ++i;
+    }
+  }
+
+  // choose best option
+  int score = INT32_MIN;
+  int retidx = -1;
+  for (int i = 0; i < runner; ++i) {
+    if (tokenLen[i] < MIN_SUFFIX) {
+      continue;
+    }
+
+    // 1. long string are likely to have less results
+    // 2. tokens at end of pattern are likely to be more relevant
+    int curScore = tokenLen[i] + i;
+
+    // iterating all children is demanding
+    if (str[tokenIdx[i] + tokenLen[i]] == '*') {
+      curScore -= 5;
+    }
+
+    // this branching is heavy
+    for (int j = tokenIdx[i]; j < tokenIdx[i] + tokenLen[i]; ++j) {
+      if (str[j] == '?') {
+        --curScore;
+      }
+    }
+
+    if (curScore >= score) {
+      score = curScore;
+      retidx = i;
+    }
+  }
+
+  return retidx;
+}
+
+int Suffix_ChooseToken_rune(const rune *str, size_t len, size_t *tokenIdx, size_t *tokenLen) {
```

Comment:
Same comments as above

Reply:
Yes. If there is no payload.
Merged two functions into one.